### PR TITLE
feat(tabbedbrowser) add option to select mode based on url pattern

### LINKED ISSUE #4960: per-domain/URL-pattern default to passthrough
I would like to set the default input mode to `passthrough` for some websites, e.g. `*.gmail.com`, `email.*.edu` and `localhost:8787` (RStudio). Reading #27 and issues linked to it and the documentations, it seems I could edit `config.py` using the command `enter-mode passthrough`, but the documentation only shows how to change *settings* based on domain/URL, but not run commands. If this is an implemented feature, could we include a few lines in the official documentation (and here)?

**Version info (see `:version`)**:
```
qutebrowser v1.7.0
Git commit: 
Backend: QtWebEngine (Chromium 65.0.3325.230)

CPython: 3.7.4
Qt: 5.11.3
PyQt: 5.11.3

sip: 4.19.18
colorama: 0.3.7
pypeg2: 2.15
jinja2: 2.10.1
pygments: 2.3.1
yaml: 3.13
cssutils: no
attr: 18.2.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebKitWidgets: yes
pdf.js: no
sqlite: 3.29.0
QtNetwork SSL: OpenSSL 1.1.1c  28 May 2019

Style: Qt5CTProxyStyle
Platform: Linux-4.19.0-5-amd64-x86_64-with-Debian-testing-buster, 64bit
Linux distribution: Debian GNU/Linux bullseye/sid (debian)
Frozen: False
Imported from /usr/lib/python3/dist-packages/qutebrowser
Using Python from /usr/bin/python3
Qt library executable path: /usr/lib/x86_64-linux-gnu/qt5/libexec, data path: /usr/share/qt5

Paths:
cache: /home/hyiltiz/.cache/qutebrowser
config: /home/hyiltiz/.config/qutebrowser
data: /home/hyiltiz/.local/share/qutebrowser
runtime: /run/user/1000/qutebrowser
system data: /usr/share/qutebrowser
```

### PR DESCRIPTION
**Pull request by AckslD**
Closes #4960

I know there is a general approach suggested in #35 and maybe this is not the desired way. But thought I would submit this if others would want to use it since I find it quite useful. I'm also open to suggested changes to the approach here.

### CODE REVIEW

**Comment by toofar** [comment]
Probably will not be included like this. If only because doing it in a slightly more general way via the extension API would probably not be tooooo much harder (although with lots more <s>arguing</s> design discussions).

A few kneejerk suggestions to look into if you/someone want's to look into it, mostly focussed on the more general case.
* I don't like the hardcoding of mode names in configdata, I see other places do it too. I don't have any suggestions about how to do it differently right now but it might be good to think about what the criteria is for a mode to be allowed or for one to be excluded. Then maybe we could tag them in modeman or something and see if that criteria could be shared with some of the other places where modes are hardcoded
* Splitting this into adding an autocmd signal into the extension API and then a mode changing component which is the first user of it might be better than this mix of general and specific in the core code.
* Lots to think about in that case about what the extension API should look like (just support url change? per tab or per window? what load states? pave the way for supporting other tab/window related events in the future? can we get the source tab/window in all cases? generic per event args?) lots of stuff to read in #35 and linked ones.
* Would be good to slice the autocmd stuff to just do one or two useful things at one point but also include some thinking about how to add more stuff in the future. Maybe just doing window events for now would be a good way to avoid the complications around including tabs in the mix.
* Can we use url patterns instead of regexes?
* No tests, probably should have been a draft PR!
* Config item description needs updating (references passthrough mode)

**Comment by AckslD** [comment]
Thanks a lot for the detailed feedback @toofar!

I could update this to use patterns and allow the execution of arbitrary commands as a tab is focused or URL change. Do you think that could make sense as a first simple `autocmd` functionality? Adding a complete `autocmd` framework with event signals and callbacks etc seems like a major effort and is probably unfortunately not something I would have time for atm. But maybe this can already add quite some useful functionalities anyway?

I'm thinking that the config entry could be something like:
```yaml
tabs.url_autocmds:
  default: {}
  type:
    name: Dict
    none_ok: true
    keytype: String
    valtype:
      name: ListOrValue
      valtype: String
  desc: Trigger commands based on URLs

    Maps url patterns to one command or list of such.
```

And this should have indeed been marked as draft, sorry about that.

**Comment by toofar** [comment]
Sorry for the delay, I keep trying to find a good chunk of time to look at how far we are from slotting something like this into the extension API. But I'm not finding it so I'll give you a less helpful response

> I could update this to use patterns and allow the execution of arbitrary commands as a tab is focused or URL change.

I think there is a couple of ways this could go:
1. added to the core code (like this PR is currently)
    * in this case perhaps the setting should use per-URL settings instead of a list? Hmm, that might be a weird fit. Like it would have to be like `:set -u site.com default_mode passthrough` and the global one would be "normal"?
2. added to the extension API (as a new hook I guess)
    * but that means thinking about how much we should put in there to make it generic-but-not-too-much effort, for instance I wouldn't try to a full generic framework with all the bells and whistles but maybe `@hook.window_event('url_changed')`, or `@hook.event(event.WindowUrlChange)` for using less strings
    * it would also mean adding modeman (or some other mode related methods) to the extension API

So I agree that (2) would be a major effort! On the other hand I don't think adding support for executing arbitrary commands is a good fit for (1). It'll add more complexity and make people want to expand on it or copy that pattern to other places and the point of the extension API is to have a place for that kind of genericness.

Another option is doing both 1 and 2! So adding it into the core like you have proposed here and then migrating it to an extension later. That brings up a few questions about how to either deprecate the config or migrate it to a plugin namespace (and let plugins register config). I guess I would want to get @The-Compiler's opinion on that (when he is done with his current responsibilities) because he probably has thoughts on that.

Either way there might be a couple of things you want to look at if you want to use this for real in it's current form:
* Isn't `_on_url_changed` called for every tab? Including background ones? Shouldn't it only apply mode changes in response to events from the active tab? There is the `cur_url_changed` signal for that.
* Is `url_changed` fired for hash changes of single page apps? Eg history.pushState()? Is that going to be problematic in some cases?

If you want to continue on with getting this PR tidied up, even with the expectation it might not be merged I have a couple of quick comments too:
* To fit with the rest of the code the new method should be called from signals (connect them in `__init__()` I think) instead of from existing handy methods. There are already signals you can use.
* Yes, URL patterns instead of regex, for consistency at least. Even if performance isn't as much of an issue here
* Tests!
* Some of the modes in the valid_values don't make sense for this feature? (for example, `yesno`)

**Comment by The-Compiler** [comment]
Some quick thoughs since I need a small break in between grading student projects anyways :sweat_smile:

> > I could update this to use patterns and allow the execution of arbitrary commands as a tab is focused or URL change.
> 
> I think there is a couple of ways this could go:
> 
> 1. added to the core code (like this PR is currently)
>    
>    * in this case perhaps the setting should use per-URL settings instead of a list? Hmm, that might be a weird fit. Like it would have to be like `:set -u site.com default_mode passthrough` and the global one would be "normal"?

I don't think it would be too weird, if we call it sommething like `input.mode_override` with `None` as default. Seems natural to me to do `:set -u example.com input.mode_override passthrough`.

> 2. added to the extension API (as a new hook I guess)
>    
>    * but that means thinking about how much we should put in there to make it generic-but-not-too-much effort, for instance I wouldn't try to a full generic framework with all the bells and whistles but maybe `@hook.window_event('url_changed')`, or `@hook.event(event.WindowUrlChange)` for using less strings
>    * it would also mean adding modeman (or some other mode related methods) to the extension API
> 
> So I agree that (2) would be a major effort! On the other hand I don't think adding support for executing arbitrary commands is a good fit for (1). It'll add more complexity and make people want to expand on it or copy that pattern to other places and the point of the extension API is to have a place for that kind of genericness.

Right. I'm not entirely sure how well this will fit in with autocmds in general, even if it seems tempting. For example, you would need an autocmd to switch to passthrough mode when going to that page, but then I suppose another one when switching away from it or something?

Given that this is a distinct use-case which people have been asking for before, and simple enough to implement, I think it might make more sense to solve this specific problem and then see if we can merge it into autocmds later, once they actually exist.

As for the extension API, that's something I'd really like someone (or a group of people) to spend a little bit of thought on how the complete API would look like before continuing to work on it, rather than just starting somewhere with no big picture in mind.

> Another option is doing both 1 and 2! So adding it into the core like you have proposed here and then migrating it to an extension later. That brings up a few questions about how to either deprecate the config or migrate it to a plugin namespace (and let plugins register config). I guess I would want to get @The-Compiler's opinion on that (when he is done with his current responsibilities) because he probably has thoughts on that.

I like the idea. In general, I'd like to move some built-in functionality to "internal" extensions in the long run, which are only allowed to use the extension API and not reach into other parts of qutebrowser's code. We mostly pulled it off for e.g. the adblocker and some utility commands (everything in `qutebrowser/components/` with a few exceptions for imports not part of `qutebrowser.api` yet).

With that in mind, it indeed becomes necessary for extensions to be able to register their own config settings, outside of `configdata.yml`. Would likely work well together with #2777 too.

> * Yes, URL patterns instead of regex, for consistency at least. Even if performance isn't as much of an issue here

Yep, makes much more sense from an user's perspective - also, regex are a poor match (hah that pun) for URLs, because e.g. `.` always needs to be escaped, and it's ugly to match on "`example.com` and it's subdomains".

**Comment by AckslD** [comment]
Thanks @toofar @The-Compiler, I started working on a new version based on your feedback and will let you know when I have something 👍

**Comment by AckslD** [comment]
In fcd95585dde68f97544f95e184ba442990b926a1 I changed to an implementation using patterns and signals. I started adding a few basic tests. The feature seems to work when I test it manually but not in the `end2end` tests for some reason. I haven't been able to figure out if I'm writing the test wrong or something else is going on. Let me know if you have any ideas :)

File: qutebrowser/config/configdata.yml [hunk 1]
```diff
@@ -1811,6 +1811,23 @@ input.match_counts:
...
+    none_ok: true
+    valid_values:
+      - normal
+      - insert
+      - hint
+      - passthrough
+      - command
+      - caret
+  supports_pattern: true
+  desc: Automatically change to the given mode when focusing on a tab.
```
Comment (mschilli87) [inline qutebrowser/config/configdata.yml:1827]: Could you rephrase this to start either with noun referring to what is selected by this option or an interrogative word?
This would make it more consistent with the rest of the config data descriptions.

```suggestion
  desc: Mode to change to when focusing on a tab.
```

Personally I'm leaning towards leaving out the `which` in this case because `null` is a valid option and thus there is not necessarily a mode to be choosen.

File: tests/end2end/features/tabs.feature [hunk 2]
```diff
@@ -1741,3 +1741,22 @@ Feature: Tab management
         And I run :undo
         And I run :message-info "Still alive!"
         Then the message "Still alive!" should be shown
+
+    Scenario: Passthrough mode override
+        When I run :set -u 1.txt input.mode_override 'passthrough'
```
Comment (toofar) [inline tests/end2end/features/tabs.feature:1746]: I suspect this URL pattern is wrong. We interpret `-u plain.string` as `*://plain.string/*` which `file:///somepath/data/numbers/1.txt` doesn't match (I assume it is opening it as an absolute path.
Try `file:///*/1.txt`

File: qutebrowser/config/configdata.yml (hunk 1 above)
Comment (AckslD) [reply qutebrowser/config/configdata.yml:1827]: Sounds good thanks!

File: tests/end2end/features/tabs.feature (hunk 2 above)
Comment (AckslD) [reply tests/end2end/features/tabs.feature]: Yep, after some debugging I also realised this, now working in 15f93546fe58748144c7ad6131c2875be5d140e1 :)

**Review by toofar — COMMENTED** [review]
Looks pretty good to me. Using the generic per-url stuff handles a lot of edge cases for you.
Congratulations on getting issue/PR number 7000!
I have a few small tweaks I would like, see inline comments.

Things I've tested:

* mode isn't overriden for pages that don't have it configured :heavy_check_mark: 
* entering passthrough, input, normal mode via the new settings works as expected when loading URLs, switching tabs, navigating back and forward, clicking through links, undoing a tab close :heavy_check_mark: 
* entering caret mode via the new setting _mostly_ works. :warning: 
* seems to play well with statusbar.show = in-mode :heavy_check_mark:
* overrides tabs.mode_on_change=persist but that is probably good :heavy_check_mark: 
* with input.insert_mode.auto_load=true it gets overriden when switching tabs but not when navigating :warning: meh, that's probably fine
* works as expected when you have multiple windows in play :heavy_check_mark: 

Regarding caret mode, I found a small issue but it seems to work in most cases so I would suggest leaving it in in this case.
Setting a caret mode override seems like it can cause the caret mode JS to be
injected too early in some cases. I get a line like this in the logs:

    js         shared:javascript_log_message:158 [userscript:_qute_js:660] Uncaught TypeError: Failed to execute 'createTreeWalker' on 'Document': parameter 1 is not of type 'Node'.

You can reproduce like:

    set -u zenpen.io input.mode_override caret 
    open -t https://zenpen.io/

It seems to enter caret mode but the status bar doesn't change colour. Not
sure if caret mode is broken or not though, I don't know how to use it. Not
sure if the JS errors are the cause of that or if there are two issues.
I think It works fine when switching tabs and navigating.

File: qutebrowser/mainwindow/tabbedbrowser.py [hunk 3]
```diff
@@ -777,6 +781,35 @@ def _on_url_changed(self, tab, url):
...
+        url = tab.url()
+        self.mode_override(url)
+
+    def mode_override(self, url: QUrl) -> None:
+        """Override mode if url matches pattern.
+
+        Args:
+            url: The QUrl to match for
+        """
+        if not url.toString():
```
[the remark this answers is not in the document]
Comment (AckslD) [reply qutebrowser/mainwindow/tabbedbrowser.py]: Cool, updated in 1b6663f24ee9f6290f6fce126a4e36bd2ca62514

File: qutebrowser/mainwindow/tabbedbrowser.py [hunk 4]
```diff
@@ -777,6 +781,35 @@ def _on_url_changed(self, tab, url):
...
+    def mode_override_on_tab_changed(self, tab):
+        """Override mode if url matches pattern on tab changed.
+
+        Args:
+            tab: The WebView where the title was changed.
+        """
+        url = tab.url()
+        self.mode_override(url)
+
+    def mode_override(self, url: QUrl) -> None:
```
[the remark this answers is not in the document]
Comment (AckslD) [reply qutebrowser/mainwindow/tabbedbrowser.py]: Sounds good, changed in 1b6663f24ee9f6290f6fce126a4e36bd2ca62514

File: qutebrowser/config/configdata.yml [hunk 5]
```diff
@@ -1811,6 +1811,23 @@ input.match_counts:
...
+  default: null
+  type:
+    name: String
+    none_ok: true
+    valid_values:
+      - normal
+      - insert
+      - hint
+      - passthrough
+      - command
```
[the remark this answers is not in the document]
Comment (AckslD) [reply qutebrowser/config/configdata.yml]: Removed in 1b6663f24ee9f6290f6fce126a4e36bd2ca62514 :+1:

**Review by toofar — APPROVED** [review]
Looks great, thanks. I'll catch up with @The-Compiler to make sure it's good to merge.

One other thing you could do if you are up for it is [clean up the commit history](https://about.gitlab.com/blog/2018/06/07/keeping-git-commit-history-clean/). There isn't much use for individual commits like "remove unused import" in the long run and it helps when looking through the commit history to be able to see a whole logical change in one commit.
In this case I think all the commits could be squashed into one. I don't know how well you know git but that would be an interactive rebase, squash all the subsequent commits onto the first one, force push.
Otherwise I'll do it when I merge.

File: qutebrowser/config/configdata.yml [hunk 6]
```diff
@@ -1811,6 +1811,21 @@ input.match_counts:
...
+input.mode_override:
+  default: null
+  type:
+    name: String
+    none_ok: true
+    valid_values:
+      - normal
+      - insert
+      - passthrough
+      - caret
```
Comment (The-Compiler) [inline qutebrowser/config/configdata.yml:1823]: Given that @toofar said this doesn't work consistently (I haven't tried myself), and I'm not sure how useful it really is, I wonder if we should support this? Easier to add it in the future if people ask for it than to take it away.

File: qutebrowser/config/configdata.yml [hunk 7]
```diff
@@ -1811,6 +1811,21 @@ input.match_counts:
...
+  supports_pattern: true
+  desc: Mode to change to when focusing on a tab.
+
+    Also when URL changes.
```
Comment (The-Compiler) [inline qutebrowser/config/configdata.yml:1827]: Maybe this could be collapsed, as in "on tab/URL changes" or so?

File: qutebrowser/mainwindow/tabbedbrowser.py [hunk 8]
```diff
@@ -777,6 +781,35 @@ def _on_url_changed(self, tab, url):
...
             self.widget.set_page_title(idx, url.toDisplayString())
 
+    def _mode_override_on_tab_changed(self, tab):
+        """Override mode if url matches pattern on tab changed.
+
+        Args:
+            tab: The WebView where the title was changed.
+        """
+        url = tab.url()
+        self._mode_override(url)
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/tabbedbrowser.py:791]: I suppose this could be a lambda where the signals are connected.

File: qutebrowser/mainwindow/tabbedbrowser.py [hunk 9]
```diff
@@ -777,6 +781,35 @@ def _on_url_changed(self, tab, url):
...
+        Args:
+            url: The QUrl to match for
+        """
+        if not url.isValid():
+            return
+        mode = config.instance.get('input.mode_override', url=url)
+        if mode:
+            log.modes.debug(
+                "Mode change to {} triggered for url {}"
+                .format(mode, url)
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/tabbedbrowser.py:805]: Could be a f-string

**Review by The-Compiler — COMMENTED** [review]
Couple of small comments, though mostly cosmetic, other than the caret one.

As for commit history, I actually prefer to only combine commits when the follow up is trivial (like the import you mentioned). For functional changes, I prefer seeing the history of how it came to be, rather than "destroying" that information for the sake of having less commits.

File: qutebrowser/config/configdata.yml (hunk 6 above)
Comment (AckslD) [reply qutebrowser/config/configdata.yml]: Fair enough, removed

File: qutebrowser/config/configdata.yml (hunk 7 above)
Comment (AckslD) [reply qutebrowser/config/configdata.yml]: Done

File: qutebrowser/mainwindow/tabbedbrowser.py (hunk 8 above)
Comment (AckslD) [reply qutebrowser/mainwindow/tabbedbrowser.py]: Done

File: qutebrowser/mainwindow/tabbedbrowser.py (hunk 9 above)
Comment (AckslD) [reply qutebrowser/mainwindow/tabbedbrowser.py]: Done

**Comment by AckslD** [comment]
Ah sorry, saw your comment too late @The-Compiler after already squashed and rebased to a single commit.

**Comment by The-Compiler** [comment]
No worries!

**Comment by toofar** [comment]
@AckslD, thanks. Sorry for all the rambling about extensions but it was a useful thought experiment for me at least.

I opened https://github.com/qutebrowser/qutebrowser/issues/7038 to track supporting caret mode. Just in case anyone asks for it at some point.

**Comment by AckslD** [comment]
No problem at all @toofar, I think it makes a lot of sense and thanks for all the feedback! :)
